nevergrad/optimization/optimizerlib.py

```diff
@@ -1815,6 +1816,139 @@ def __init__(
 )
 
 
+class _BayesOptim(base.Optimizer):
+    def __init__(
+        self,
+        parametrization: IntOrParameter,
+        budget: tp.Optional[int] = None,
+        num_workers: int = 1,
+        *,
+        init_budget: tp.Optional[int] = None,
+        pca: tp.Optional[bool] = False,
+        n_components: tp.Optional[float] = 0.95,
+        prop_doe_factor: tp.Optional[float] = None,
+    ) -> None:
+        super().__init__(parametrization, budget=budget, num_workers=num_workers)
+
+        from bayes_optim.extension import PCABO as PcaBO
+
+        # from bayes_optim.extension import RealSpace
+        from bayes_optim import BO as BayesOptimBO
+        from bayes_optim import RealSpace
+        from bayes_optim.surrogate import GaussianProcess
+
+        lb, ub = 1e-7 - np.pi / 2, np.pi / 2 - 1e-7
+        space = RealSpace([lb, ub]) * self.dimension
+
+        # hyperparameters of the GPR model when pca = False
+        thetaL = 1e-10 * (ub - lb) * np.ones(self.dimension)
+        thetaU = 10 * (ub - lb) * np.ones(self.dimension)
+        model = GaussianProcess(thetaL=thetaL, thetaU=thetaU)  # create the GPR model
+
+        self._buffer: tp.List[float] = []
+        self._newX: tp.List[float] = []
+        self._losses: tp.List[float] = []
+
+        # Setting DoE size as a percentage of the total budget if prop_doe_factor is not None
+        if prop_doe_factor and budget:
+            init_budget = round(prop_doe_factor * budget)
+
+        if pca:
+            self._alg = PcaBO(
+                search_space=space,
+                obj_fun=None,  # Assuming that this is not used :-)
+                DoE_size=init_budget if init_budget is not None else 5,
+                max_FEs=budget,
+                verbose=True,
+                n_point=1,  # We start with a sequential procedure, maybe we'll extend in a second moment
+                n_components=n_components,
+                acquisition_optimization={"optimizer": "BFGS"},
+            )
+        else:
+            self._alg = BayesOptimBO(
+                search_space=space,
+                obj_fun=None,  # Assuming that this is not used :-)
+                model=model,
+                DoE_size=init_budget if init_budget is not None else 5,
+                max_FEs=budget,
+                verbose=True,
+            )
+
+    def _internal_ask_candidate(self) -> p.Parameter:
+        if not self._buffer:
+            self._newX = []
+            self._losses = []
+            candidate = self._alg.ask()
+            if not isinstance(candidate, list):
+                candidate = candidate.tolist()
+            self._buffer = candidate
+        x_probe = self._buffer.pop()
+        data = np.tan(np.array(x_probe, copy=False))
+        candidate = self.parametrization.spawn_child().set_standardized_data(data)
+        candidate._meta["x_probe"] = x_probe
+        self._newX.append(candidate._meta["x_probe"])
```

Comment:
please move this line to `_internal_tell_candidate`. This should add robustness and clarity by filling losses and _newX together.

Reply:
candidate is returned by `_internal_ask_candidate`, don't we need to apply `tan` and `set_standardized_data` before evaluation of the candidate?

Reply:
I don't understand how your question is related to this line (`self._newX.append(candidate._meta["x_probe"])`)
This will do just the same in tell, storing the value of `x_probe` as it is, the way you normalized it.

Still, the question is interesting, I think there is something fishy here. What is the domain of the "ask" in bayes_optim? if you apply tan on it, it should be -pi/2 to pi/2 to map it to R, I don't think it is

Reply:
For the transform, let's use this:
https://github.com/facebookresearch/nevergrad/blob/fdb364181e9bd3a2d68ad75e1a7ae0605e7ed8f4/nevergrad/parametrization/transforms.py#L218-L244

If your domain is [0, 1], then create `self._transform = ArctanBound(0, 1)`, and `self._transform.backward(data)` will map from [0, 1] to R (and `self._transform.backward` will do the other way round)